### PR TITLE
Adding MigrationHelper

### PR DESCRIPTION
**Pull request by franchester4**
Closes: https://snapsheettech.atlassian.net/browse/SNAPTX-3899

Adds migration functionality to the gem with class `SsmConfig::MigrationHelper`. Supports `migrate` and `unmigrate` methods.

`migrate` implemented by taking the hash and traversing through it to insert each value into the table (DFS). In the case that a validation is violated, table will be reverted to initial state

`unmigrate` removes all rows with matching file name.

Added specs, tested migrations in snaptx

Also a minor restructuring in the rspecs

### CODE REVIEW

File: lib/ssm_config/migration_helper.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    def migrate
+      added = []
+      keys_hash = accessor_key_hash(hash) # starting layer is always hash
+      last = nil
+      keys_hash.each do |accessor_key, value|
+        last = accessor_key
+        added.push(@model_name.constantize.create!(:file => @file_name, :accessor_keys => accessor_key, :value => value.to_s, :datatype => determine_class(value)))
+      end
+    rescue ActiveRecord::RecordInvalid => e
+      Rails.logger.error("#{e.message} was raised because of faulty data with accessor_key #{last}")
```
Comment (lorman) [inline lib/ssm_config/migration_helper.rb:17]: Add a spec for this. You can probably to it with `allow(model_name).to receive(create!).and raise` or something like that.

**Review by lorman — APPROVED** [review]
NIT: small request to add spec. Also would it make sense to name the methods in the migration helper `up` and `down` similar to the rails migration methods?

**Review by holmes2 — APPROVED** [review]
minor changes to README.. open to suggestions.. otherwise LG

File: lib/ssm_config/migration_helper.rb (hunk 1 above)
Comment (franchester4) [reply lib/ssm_config/migration_helper.rb:17]: Got it. I wasn't able to figure it out earlier and reasoned its more of an ActiveRecord problem, but I'll look into it more

File: spec/lib/ssm_config/migration_helper_spec.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    context 'when a more complex file exists' do
+      it 'migrates file correctly' do
+        migration_helper.up
+        expect(db_query.hash).to eq(yml_query.hash)
+      end
+    end
+
+    context 'when ActiveRecord::RecordInvalid is raised' do
```
Comment (lorman) [inline spec/lib/ssm_config/migration_helper_spec.rb:40]: 🎉

**Review by lorman — APPROVED** [review]
CR: LG

**Review by lorman — CHANGES_REQUESTED** [review]
Let's bump the gem version.

File: lib/ssm_config.rb [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
 require './lib/ssm_config/ssm_storage/empty.rb'
+require './lib/ssm_config/errors.rb'
+require './lib/ssm_config/migration_helper.rb'
 require 'active_support/core_ext/hash/indifferent_access'
 require 'active_support/time'
-require './lib/ssm_config/errors.rb'
 
 module SsmConfig
-  VERSION = '1.1.0'.freeze
+  VERSION = '1.2.0'.freeze
```
Comment (lorman) [inline lib/ssm_config.rb:10]: Sorry, didn't see this

**Review by lorman — APPROVED** [review]

**Pull request merged**